### PR TITLE
Add guest-stats to VMI with dirty-rate average and variance, important for live-migrations

### PR DESCRIPTION
**Pull request by iholder101**
**What this PR does / why we need it**:
One of the most important key factors for a live-migration's success is the VM's dirty rate. For example, if the dirty rate is higher than the network bandwidth the VM wouldn't ever converge, IOW the migration would never finish.

This PR exposes VMI's dirty rate's average and variance.

See https://github.com/kubevirt/kubevirt/issues/10301 for more info.

**Implementation summary**:
- Cache `GetDomainStats()` call on the virt-launcher's.
- Adds a GuestStats struct to VMI's status, currently containing only dirty-rate related statistics.
- Every minute, update the statistics for all VMIs.

**Release note**:

```release-note
Add guest-stats to VMI with dirty-rate average and variance, important for live-migrations
```

-----------

**Usage example**:
Let's create a new VMI and make sure the dirty rate statistics are being updated:
```bash
> k create -f examples/vmi-fedora.yaml 
selecting docker as container runtime
virtualmachineinstance.kubevirt.io/vmi-fedora created

> k get vmi -w -o yaml | grep dirty -i -A3
selecting docker as container runtime
    dirtyRate:
      average: 8
      sampleCount: 1
--
    dirtyRate:
      average: 4
      sampleCount: 2
      variance: 16
--
    dirtyRate:
      average: 2.6666666666666665
      sampleCount: 3
      variance: 14.222222222222221
```

As can be seen the dirty rate is being updated.
Now, let's stress vmi-fedora and see what happens to the dirty rate:
```bash
> v console vmi-fedora
selecting docker as container runtime
Successfully connected to vmi-fedora console. The escape sequence is ^]

vmi-fedora login: fedora
Password: 
[fedora@vmi-fedora ~]$ stress-ng --vm-bytes 512M --vm-keep -m 1
stress-ng: info:  [825] defaulting to a 86400 second (1 day, 0.00 secs) run per stressor
stress-ng: info:  [825] dispatching hogs: 1 vm
```

And then we see:
```bash
--
    dirtyRate:
      average: 2
      sampleCount: 6
      variance: 9.333333333333336
--
    dirtyRate:
      average: 79.14285714285714
      sampleCount: 7
      variance: 35714.12244897959
--
    dirtyRate:
      average: 129.75
      sampleCount: 8
      variance: 49177.4375
```

As expected, the dirty rate average grows significantly.

### CODE REVIEW

**Comment by rmohr** [comment]
Is this something which can be seen over metrics? My main reservation on something like this may be status update frequency on bigger scale. Could you include exact behaviour on that regard?

**Comment by rmohr** [comment]
> Is this something which can be seen over metrics? My main reservation on something like this may be status update frequency on bigger scale. Could you include exact behaviour on that regard?

Reading helps. 1 minute.

**Comment by iholder101** [comment]
> > Is this something which can be seen over metrics? My main reservation on something like this may be status update frequency on bigger scale. Could you include exact behaviour on that regard?
> 
> Reading helps. 1 minute.

:)

The reasoning here is that metrics that should be used in-code (e.g. in the future can be used by virt-controller in order to asses how to configure migrations) better not rely on Prometheus, which can be disabled and is not intended for this kind of use.

w.r.t. scale: this information is needed for the long-term and we don't need it to be very accurate. In other words, I don't mind even lowering the sample period from 1 to 5 minutes. Also bear in mind that the guest-stats-controller does not define any event handler nor any queues. In other word, it only wakes up once in that period (1 minute as per the current implementation) - if something fails the stats won't be updated until the next period exceeds.

File: pkg/virt-handler/guest-stats-updater.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+		launcherClient:       virtcache.LauncherClientInfoByVMI{},
+		podIsolationDetector: isolation.NewSocketBasedIsolationDetector(virtShareDir),
+		logger:               log.Logger("guest-stat-controller").With("node", host),
+	}
+}
+
+func (c *GuestStatsController) Run(stopCh <-chan struct{}) error {
+	cache.WaitForCacheSync(stopCh, c.vmiInformer.HasSynced)
+
+	go wait.Until(c.execute, statsUpdateInterval, stopCh)
```
Comment (rmohr) [inline pkg/virt-handler/guest-stats-updater.go:57]: I think you can also lower this to 15 or 30 seconds if you do a conceptial change. I would recommend one conceptual change:

Wait for each VM individually "roughly" for the update period with randomization, to distribute the updates of all VMs over the period.
 
 If we don't do that, we end up with hammering on a bigger node the API with 300 requests in one go, then  one minute nothing, then we hammer on the apiserver again.
 
 250 VMs/15 is just ~17 requests per second, compared to one peak of 300 requests which will most certainly hit the rate-limiter or hurt the apiserver performance. 30 seconds would be  just 8 requests per seconds, which is definitely not a concern.

**Review by iholder101 — COMMENTED** [review]
Some dilemmas I have @vladikr @rmohr

File: pkg/virt-handler/guest-stats-updater.go (hunk 1 above)
Comment (iholder101) [reply pkg/virt-handler/guest-stats-updater.go:57]: Interesting!
Sounds like a great improvement to me. Thank you @rmohr!

**Comment by iholder101** [comment]
Also: do you think that we should track these statistics on the VM level instead of the VMI?

On the one hand it will cause the stats to be relevant between reboots.
On the other hand, I'm not sure that's desired (VM can behave differently after boot for some reason). Also, sometimes VMI without a backing VM need to migrate.

**Comment by rmohr** [comment]
> Also: do you think that we should track these statistics on the VM level instead of the VMI?
> 
> On the one hand it will cause the stats to be relevant between reboots. On the other hand, I'm not sure that's desired (VM can behave differently after boot for some reason). Also, sometimes VMI without a backing VM need to migrate.

I think that metrics are there for historic investigations. It is a typical thing which maybe get spropagated to the VM level, but VMI is the right layer where it should be used to maybe perform controller-triggered optimizations.

**Comment by iholder101** [comment]
@rmohr Thanks for your great suggestions! PTAL
@vladikr FYI

Also - are we concerned regarding the performance penalty on the guest? IIUC the penalty is pretty low, especially with the page sampling method and only 1 second of sampling (guest memory isn't being locked), however it's still somewhat of a penalty.

Do we want to exclude high-performance VMs automatically? Add a Kubevirt CR knob? Opt-out by an annotation on a VM-basis level?

**Comment by xpivarc** [comment]
Few questions and thoughts:

1. I think the cost is still high even if we offset it as @rmohr suggested. 
2. Could we squash regular updates with statistic updates?
3.  Are we sure we will need these statistics exposed on the VMI?
4. Could we make the decision/auto-configuration locally, on a node? Removing the need to expose the statistics.
5.  Do we need to update regularly the statistics or do we need the only one at the decision point?
6. How will these statistics help us now?

What I am missing here is a higher-level view of how should this thing work in the future.

**Comment by rthallisey** [comment]
>Implementation summary:
>
>-  Cache GetDomainStats() call on the virt-launcher's.
>-  Adds a GuestStats struct to VMI's status, currently containing only dirty-rate related statistics.
>-  Every minute, update the statistics for all VMIs.
>
Putting something in every vmi's status field and updating it every minute or so it pretty significant. 
Can you describe the use-cases in more detail please?

**Comment by rthallisey** [comment]
I just noticed there's more detail on the issue you shared.
>For an advanced user, or even a Kubevirt developer, it wouldn't be trivial to set these configurations correctly. The problem only gets worse if the configurations need to adapt to the VM's dynamic behavior (e.g. is the guest under pressure right now? Is there a large probability of it getting pressured soon?).
>
Maybe you can speak to the value-add for these affected users.  For example, how often would the advanced user/kubevirt developer perform live-migration?   I think that would help me, at least, understand better where you're coming from.

**Comment by iholder101** [comment]
Hey @xpivarc! Thanks for having a look.

> Few questions and thoughts:
> 
> 1. I think the cost is still high even if we offset it as @rmohr suggested.

Which cost are you referring to? The cost to the API-server/etcd or the guest?

While the penalty is there, it's supposed to be relatively small with the page sampling method since the guest pages aren't being locked. Perhaps we can ask libvirt experts to chime in here.

In any case, I'm open to address this concern (for both API server and the guest) by making the sample interval larger. We can sample every 5 minutes even, perhaps even more if necessary. We only care about the long-term average here and I think we can agree that with a large enough sample period the performance concern becomes pretty negligible. In addition, I'm opened to an opt-out mechanism of some sort as said in [the comment above](https://github.com/kubevirt/kubevirt/pull/10302#issuecomment-1686167669).

> 2. Could we squash regular updates with statistic updates?

I think that being bound to other updates is problematic, partly because of the performance concern. Regular controllers tend to reenqueue VMIs when being updated, but here we only act periodically, and as said above we might want that sampling period to be quite large.

> 3. Are we sure we will need these statistics exposed on the VMI?
> 6. How will these statistics help us now?

Combining these the two questions:
I think that in general dirty rate is one of the most important statistics for migratable VMIs. Especially now, when we don't support auto migration configs, we want to ease the configuration of migrations. Without knowing the dirty rate it's much more probable to mis-configure migrations which might lead to non-converging VMs. So IMHO I think this information is very valuable, even (and perhaps, especially) if we don't already support auto-configuring migrations.

> 4. Could we make the decision/auto-configuration locally, on a node? Removing the need to expose the statistics.

Further to the above, I think that a technical problem with your suggestion is virt-handler restarts, which might happen. If virt-handler would keep this information solely in-memory, it all would be gone if virt-handler restarts for any reason.

> 5. Do we need to update regularly the statistics or do we need the only one at the decision point?

The way it works from the libvirt side is that you can calculate the dirty rate in the last 1<=X<=60 seconds. Only sampling that small time-frame is not accurate enough, so it's important to keep a running-average of the dirty rate over time.

However, this dirty rate average would only be needed once the VM is migrating.

> What I am missing here is a higher-level view of how should this thing work in the future.

Actually, I think we're moving ahead cautiously on purpose.
What I mean is that before we finalize or even begin discussing the auto-migration-config APIs and such, we want to first expose some building blocks. This would both grant us the opportunity to see how difficult it is to migrate VMs with a certain dirty rate while using certain migration techniques (e.g. multifd, post-copy, etc), and as a bi-product let the VM owners know their workloads better, i.e. expose the needed statistics to reduce the probability of misconfiguring migrations.

**Comment by rmohr** [comment]
> Few questions and thoughts:
> 
> 1. I think the cost is still high even if we offset it as @rmohr suggested.
> 2. Could we squash regular updates with statistic updates?
> 3. Are we sure we will need these statistics exposed on the VMI?
> 4. Could we make the decision/auto-configuration locally, on a node? Removing the need to expose the statistics.
> 5. Do we need to update regularly the statistics or do we need the only one at the decision point?
> 6. How will these statistics help us now?
> 
> What I am missing here is a higher-level view of how should this thing work in the future.

I also want to add, that for instance HPO (horizontal pod autoscaler) is also taking its metrics for decision making from the metrics API and not from the objects themselves. It may still be an option, despite my suggestions on how to minimize the API impact.

**Comment by rthallisey** [comment]
Thanks @iholder101, this is helpful.

>What I meant is that even for advanced users or kubevirt developers, that have a lot of knowledge about the mechanics of live-migrations, it would be hard to configure migrations. Therefore, it would obviously be hard for non-advanced users as well.
>
I'm not going to dispute your assertion, however I'm not yet convinced that it's worth extending the kubevirt api space to make live-migration easier.  Maintaining anything on the api is expensive, so I'm wondering specific what the ROI is.

Since live-migration is difficult, would a tool be helpful?  For instance if we reported this data as metrics, a tool can scrape Prometheus and present the data in a useful way.  Is that avenue possible?

**Comment by iholder101** [comment]
Thanks everyone for your feedback.
@vladikr WDYT?

**Comment by vladikr** [comment]
Hey,

I'm late to review this PR, sorry.

Itamar and I have spoken offline. We agreed to separate the StartDrityRateCalculation from the collection of domain stats to reduce the dependency between the two processes.
@iholder101 perhaps you can expand on this.

**Comment by rmohr** [comment]
> @rmohr that's very interesting. I'm not sure how can we expose the dirty rate there.

I think it would have to be done with custom or external metrics (https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#support-for-metrics-apis).

**Comment by vladikr** [comment]
> Thanks @iholder101, this is helpful.
> 
> > What I meant is that even for advanced users or kubevirt developers, that have a lot of knowledge about the mechanics of live-migrations, it would be hard to configure migrations. Therefore, it would obviously be hard for non-advanced users as well.
> 
> I'm not going to dispute your assertion, however I'm not yet convinced that it's worth extending the kubevirt api space to make live-migration easier. Maintaining anything on the api is expensive, so I'm wondering specific what the ROI is.
> 
> Since live-migration is difficult, would a tool be helpful? For instance if we reported this data as metrics, a tool can scrape Prometheus and present the data in a useful way. Is that avenue possible?

Thanks @rthallisey . 
I think that looking at an individual live migration won't reveal the extent of the issue. 
However, as you know, we update KubeVirt via live migration. VMs that cannot migrate in a reasonable timeframe will delay or prevent upgrades. 
Cluster admins that have a specific window for maintenance suffer greatly from a lack of optimization in this area.

**Comment by iholder101** [comment]
I've reworked the PR to address the comments raised here. The main changes are:
1. Now virt-launcher keeps the statistics internally.
2. virt-handler would, every now and then, query the currently measured stats. The stats would be updated to the VMI's status if one of the following is true:
   * It's the first sample.
   * More than 150 samples had been measured from last update.
   * The new dirty-rate average is different in +-10%.
4. This feature is now guarded with a feature gate that is off by default.

We can of course discuss the numbers here, but the idea is that for the vast majority of workloads the average dirty rate won't change drastically all the time, and therefore would be updated just once in a while, therefore easing on the API server significantly. In addition, as time goes by it's naturally harder to change the average.

There are for sure improvements to be done here. For example, it would be better if virt-handler wouldn't poll the guest stats from the launcher, and instead just waited to be notified by the launcher. This is supported by grpc with `stream`s, although it has a [known issue](https://github.com/golang/mock/pull/163) when working with mockgen (which is [not maintained anymore](https://github.com/kubevirt/kubevirt/issues/10355)). Also, we should keep the average relevant to the last X samples, so that it would be (almost) impossible to affect it. In addition, I am willing to consider the metrics approach when moving forward.

Saying that, this is an alpha version that's supposed to grant us the "feel" of this feature before we make too large design decisions. As with Alpha, I prefer to keep it as simple and easy to change as possible. Since there are no APIs here pretty much everything else is an implementation details that could always be improved.

Let me know what you think.
FYI @vladikr 

/test pull-kubevirt-build 
/test pull-kubevirt-e2e-k8s-1.26-sig-compute
/test pull-kubevirt-e2e-k8s-1.27-sig-compute
/test pull-kubevirt-e2e-k8s-1.27-sig-compute-migrations

**Comment by xpivarc** [comment]
The latest version makes sense, keeping it behind the feature gate enables us to experiment and validate the idea and performance. It would be great if we could also consider alternatives that would not require exposing the dirty rate on each VM.

**Pull request closed without merging**